lib/mix/lib/mix/compilers/application_tracer.ex

```diff
@@ -57,7 +57,7 @@ defmodule Mix.Compilers.ApplicationTracer do
           # If the module belongs to this application (from another compiler), we skip it.
           # If the module is excluded, we skip it too.
           with path when is_list(path) <- :code.which(module),
-               [module_app, "ebin", _beam] <- path |> Path.split() |> Enum.take(-3),
+               module_app = app(path),
```

Comment:
replacing `<-` with `=` makes this less forgiving. We'd also incorrectly handle when the app name legitimately contains a dash. I don't think either is a concern but happy to change it!

Reply:
Yeah, i don't want compilation to break because we can't emit an optional warning, so I think `<-` is a better call.